File: diagflow/tools/v3tools.py
"""
v3 Tool definitions — plain async functions wrapped as Anthropic SDK tools.

Each tool is a ToolDef(name, description, input_schema, handler).
No Tool class boilerplate — just async functions with JSON Schema params.
"""

from __future__ import annotations

import asyncio
import hashlib
import hmac
import logging
import os
import time
from typing import Any

import httpx
import paramiko
import yaml
from pathlib import Path

from diagflow.core.diag_agent import ToolDef
from diagflow.core.memory import EvidencePool

logger = logging.getLogger(__name__)


# ===========================================================================
# Tool 1: query_yarn — direct YARN RM query
# ===========================================================================

async def _query_yarn_handler(
    cluster,
    action: str = "list_apps",
    app_id: str = "",
    app_name: str = "",
) -> str:
    await cluster._ensure_node_data()
    master = cluster._find_node("master1")
    if not master:
        return "Error: No master node found"
    ipv6 = master.get("ipv6", "")
    rm_url = f"http://[{ipv6}]:23188/ws/v1/cluster"

    async with httpx.AsyncClient(timeout=10) as c:
        if action == "list_apps":
            states = app_name or "RUNNING,FAILED,KILLED,FINISHED"
            resp = await c.get(f"{rm_url}/apps?states={states}")
            data = resp.json()
            apps = data.get("apps", {}).get("app", [])
            if not apps:
                return "0 YARN apps found."
            result = []
            for a in apps[:20]:
                fs = a.get("finalStatus", a.get("state", ""))
                diag = (a.get("diagnostics", "") or "")[:200]
                result.append(f"  {a.get('name','?')}  id={a.get('id','?')}  state={a.get('state','?')}  final={fs}")
                if fs == "FAILED" and diag:
                    result.append(f"    diagnostics: {diag}")
            failed = sum(1 for a in apps if a.get("finalStatus") == "FAILED")
            return f"{len(apps)} YARN apps ({failed} FAILED):\n" + "\n".join(result)

        elif action == "app_nodes":
            resp = await c.get(f"{rm_url}/apps/{app_id}/appattempts")
            data = resp.json()
            attempts = data.get("appAttempts", {}).get("appAttempt", [])
            nodes = list({a.get("host", "") for a in attempts if a.get("host")})
            container_ids = [a.get("containerId", "") for a in attempts if a.get("containerId")]

            lines = [f"App {app_id} on {len(nodes)} nodes:"]
            for n in nodes:
                lines.append(f"  node: {n}")

            if container_ids:
                lines.append(f"  containers: {', '.join(container_ids[:10])}")

                # Build container log directory paths (production format):
                # /data/yarn/logs/{app_id}/{container_id}/taskmanager.err
                lines.append("")
                lines.append("  === Two-tier log retrieval ===")
                lines.append("")
                lines.append("  Tier 1 — yarn CLI (run on master node):")
                lines.append(f"    yarn logs -applicationId {app_id} -logFiles stderr -size -1 | tail -200")
                lines.append(f"    yarn logs -applicationId {app_id} -logFiles stdout -size -1 | tail -200")
                lines.append(f"    # Add -containerId {container_ids[0]} to target a specific container")
                lines.append("")
                lines.append("  Tier 2 — direct file access (if yarn logs fails, SSH to the node):")
                for cid in container_ids[:5]:
                    # containerId: container_e06_1783421408688_12571_02_000003
                    log_dir = f"/data/yarn/logs/{app_id}/{cid}"
                    # Find the node for this container
                    c_node = attempts[container_ids.index(cid)].get("host", "?") if cid in container_ids else "?"
                    lines.append(f"    # Container {cid} on {c_node}")
                    lines.append(f"    ssh_exec(node='{c_node}', cmd='ls {log_dir}/taskmanager.* 2>/dev/null')")
                    lines.append(f"    ssh_exec(node='{c_node}', cmd='tail -200 {log_dir}/taskmanager.err 2>/dev/null')")
            else:
                lines.append("  No container info available.")

            return "\n".join(lines)

        return f"Unknown action: {action}"


def make_query_yarn_tool(cluster):
    return ToolDef(
        name="query_yarn",
        description="Query YARN RM. Use 'list_apps' to see all apps (FAILED first). Use 'app_nodes' with app_id to find nodes.",
        input_schema={
            "type": "object",
            "properties": {
                "action": {"type": "string", "description": "'list_apps' or 'app_nodes'"},
                "app_id": {"type": "string", "description": "YARN app ID for 'app_nodes'"},
                "app_name": {"type": "string", "description": "Comma-separated states: RUNNING,FAILED,KILLED"},
            },
            "required": ["action"],
        },
        handler=lambda **kw: _query_yarn_handler(cluster, **kw),
    )


# ===========================================================================
# Utilities
# ===========================================================================


def _validate_command(cmd: str) -> tuple[bool, str]:
    """Validate that a shell command is read-only and safe.

    Returns (allowed, reason).
    """
    from diagflow.config import get_config
    cfg = get_config()

    cmd_clean = cmd.strip()
    cmd_lower = cmd_clean.lower()

    # Check forbidden patterns
    for forbidden in cfg.security.ssh_forbidden:
        if forbidden in cmd_lower:
            return False, f"Forbidden pattern '{forbidden}' in command"

    # Check allowed first word
    first_word = cmd_lower.split()[0] if cmd_lower.split() else ""
    if first_word in cfg.security.ssh_allowed:
        return True, ""
    # Also allow piped commands starting with allowed words
    if "|" in cmd_clean:
        first_in_pipe = cmd_lower.split("|")[0].strip().split()[0] if cmd_lower.split("|")[0].strip().split() else ""
        if first_in_pipe in cfg.security.ssh_allowed:
            return True, ""

    return False, f"Command '{first_word}' not in read-only allowlist"


# ===========================================================================
# Tool 2: ssh_exec — SSH to cluster node and run command
# ===========================================================================

async def _ssh_exec_handler(cluster, node_name: str, cmd: str, timeout_s: int = 15) -> str:
    # ---- Command safety check ----
    allowed, reason = _validate_command(cmd)
    if not allowed:
        logger.warning("Blocked unsafe SSH command: %s (reason: %s)", cmd[:100], reason)
        return f"Command blocked by safety filter: {reason}"

    await cluster._ensure_node_data()
    node = cluster._find_node(node_name)
    if not node:
        return f"Error: Node '{node_name}' not found"
    ipv6 = node.get("ipv6", "")
    user = node.get("access_user", "root")
    passwd_raw = node.get("passwd", "")

    # Decode base64 password
    import base64
    try:
        passwd = base64.b64decode(passwd_raw).decode("utf-8")
    except Exception:
        passwd = passwd_raw

    if not ipv6:
        return f"Error: No IPv6 for '{node_name}'"

    loop = asyncio.get_event_loop()
    client = paramiko.SSHClient()
    client.set_missing_host_key_policy(paramiko.AutoAddPolicy())

    def _ssh():
        client.connect(hostname=ipv6, port=22, username=user, password=passwd,
                       timeout=timeout_s, look_for_keys=False, allow_agent=False)
        stdin, stdout, stderr = client.exec_command(cmd, timeout=timeout_s)
        out = stdout.read().decode("utf-8", errors="replace")
        err = stderr.read().decode("utf-8", errors="replace")
        return (out, err)

    try:
        out, err = await loop.run_in_executor(None, _ssh)
        result = out.strip() or err.strip()
        if not result:
            result = "(empty)"
        return result
    except Exception as e:
        return f"SSH failed: {e}"
    finally:
        try:
            client.close()
        except Exception:
            pass


def make_ssh_exec_tool(cluster):
    return ToolDef(
        name="ssh_exec",
        description="Execute read-only shell commands on cluster nodes via SSH. Find logs, grep errors, check processes.",
        input_schema={
            "type": "object",
            "properties": {
                "node_name": {"type": "string", "description": "Node: 'master1', 'core1', or full node_name"},
                "cmd": {"type": "string", "description": "Shell command: find, grep, tail, ps, curl, df, free"},
                "timeout_s": {"type": "integer", "description": "SSH timeout in seconds", "default": 15},
            },
            "required": ["node_name", "cmd"],
        },
        handler=lambda **kw: _ssh_exec_handler(cluster, **kw),
    )


# ===========================================================================
# Tool 3: deepwiki_query — MCP-based known-issue lookup
# ===========================================================================


async def _deepwiki_handler(component: str, question: str, version: str = "") -> str:
    from diagflow.config import get_config
    cfg = get_config()
    repo = cfg.components.repo_map.get(component.lower())
    if not repo:
        return f"Unknown component '{component}'. Supported: {list(cfg.components.repo_map.keys())}"

    from mcp import ClientSession
    from mcp.client.streamable_http import streamablehttp_client

    try:
        async with streamablehttp_client("https://mcp.deepwiki.com/mcp") as (read, write, _):
            async with ClientSession(read, write) as session:
                await session.initialize()
                result = await session.call_tool(
                    "ask_question",
                    {"repoName": repo, "question": f"{question} (version: {version})" if version else question},
                )
                parts = [getattr(b, "text", "") for b in result.content if getattr(b, "text", None)]
                return f"[{repo}]\n" + "\n".join(parts) if parts else "(empty)"
    except Exception as e:
        return f"DeepWiki error: {e}"


def make_deepwiki_tool():
    return ToolDef(
        name="deepwiki_query",
        description="Query known issues in open-source repos. Component→repo: flink→apache/flink, hdfs→apache/hadoop, kafka→apache/kafka.",
        input_schema={
            "type": "object",
            "properties": {
                "component": {"type": "string", "description": "Component: flink, hdfs, yarn, kafka, spark, hbase"},
                "question": {"type": "string", "description": "Natural language question about known bugs"},
                "version": {"type": "string", "description": "Component version"},
            },
            "required": ["component", "question"],
        },
        handler=_deepwiki_handler,
    )


# ===========================================================================
# Tool 4: fingerprint_match
# ===========================================================================

async def _fingerprint_handler(component: str, error_pattern: str, version: str = "") -> str:
    fp = hashlib.md5(f"{component}:{error_pattern}:{version}".encode()).hexdigest()[:16]
    # KB is injected per-diagnosis; fallback if not set
    return f"No fingerprint match for {component}/{error_pattern} (fp={fp}). Proceed with diagnosis."


def make_fingerprint_tool():
    return ToolDef(
        name="fingerprint_match",
        description="Check if this error matches a known case. Fast-path before full diagnosis.",
        input_schema={
            "type": "object",
            "properties": {
                "component": {"type": "string", "description": "Component: flink, hdfs, yarn"},
                "error_pattern": {"type": "string", "description": "Key error: OutOfMemoryError, CheckpointExpired..."},
                "version": {"type": "string", "description": "Component version"},
            },
            "required": ["component", "error_pattern"],
        },
        handler=_fingerprint_handler,
    )


# ===========================================================================
# Tool 5: umrAgent proxy (for CheckProcess / GetBaseInfo / GetAppList)
# ===========================================================================

async def _umr_agent_handler(
    cluster, node_name: str, action: str, params: dict | None = None,
) -> str:
    await cluster._ensure_node_data()
    node = cluster._find_node(node_name)
    if not node:
        return f"Error: Node '{node_name}' not found"
    ipv6 = node.get("ipv6", "")
    agent_key = node.get("agent_key", "") or node.get("umr_agent_key", "")
    if not ipv6 or not agent_key:
        return f"Error: No ipv6/agent_key for '{node_name}'"

    all_params = {"Action": action, "Date": str(int(time.time() * 1000))}
    if params:
        all_params.update({k: str(v) for k, v in params.items()})

    # HMAC-SHA1 signing (replicates util/agent.js)
    string_to_sign = "".join(all_params[k] for k in sorted(all_params))
    sig = hmac.new(agent_key.encode(), string_to_sign.encode(), hashlib.sha1).hexdigest()
    all_params["Signature"] = sig

    query = "&".join(f"{k}={v}" for k, v in all_params.items())
    url = f"http://[{ipv6}]:65431/?{query}"

    async with httpx.AsyncClient(timeout=15) as c:
        resp = await c.get(url)
        return resp.text


def make_umr_agent_tool(cluster):
    return ToolDef(
        name="call_umr_agent",
        description="Call umrAgent on cluster nodes. Actions: GetLogs(Path,Keywords,MaxLines,Since), CheckProcess(ProcessName), GetBaseInfo, GetAppList.",
        input_schema={
            "type": "object",
            "properties": {
                "node_name": {"type": "string", "description": "Node: 'master1', 'core1', or full node_name"},
                "action": {"type": "string", "description": "umrAgent Action: GetLogs, CheckProcess, GetBaseInfo, GetAppList"},
                "params": {"type": "object", "description": "Action-specific params as JSON"},
            },
            "required": ["node_name", "action"],
        },
        handler=lambda **kw: _umr_agent_handler(cluster, **kw),
    )


# ===========================================================================
# Factory: build all tools for a cluster
# ===========================================================================

def build_v3_tools(cluster, kb=None) -> list[ToolDef]:
    tools = [
        make_query_yarn_tool(cluster),
        make_ssh_exec_tool(cluster),
        make_umr_agent_tool(cluster),
        make_deepwiki_tool(),
        make_fingerprint_tool(),
    ]
    # Inject KB into fingerprint tool
    if kb:
        fp_tool = tools[-1]
        async def _kb_handler(**kw):
            hit = kb.fingerprint_match(
                kw.get("component", ""), kw.get("error_pattern", ""), kw.get("version", "")
            )
            if hit:
                return f"✅ Known issue (confidence: high):\n  Root cause: {hit.get('root_cause','?')}\n  Suggestions: {hit.get('suggestions',[])}"
            return await _fingerprint_handler(**kw)
        tools[-1] = ToolDef(
            name="fingerprint_match",
            description=fp_tool.description,
            input_schema=fp_tool.input_schema,
            handler=_kb_handler,
        )
    return tools

File: internal/tools/v3tools/v3tools.go
// Package v3tools implements the five v3 diagnostic tools.
package v3tools

import (
	"context"
	"crypto/hmac"
	"crypto/md5"
	"crypto/sha1"
	"encoding/hex"
	"fmt"
	"sort"
	"strings"
	"time"

	"github.com/Admaing/diagflow/internal/cluster"
	"github.com/Admaing/diagflow/internal/tools"
)

// BuildV3Tools returns the five v3 tools, injecting a KB-backed fingerprint
// handler when kb is non-nil.
func BuildV3Tools(c cluster.Cluster, kb fingerprintMatcher) []tools.Def {
	list := []tools.Def{
		queryYarnTool(c),
		sshExecTool(c),
		umrAgentTool(c),
		deepwikiTool(),
		fingerprintTool(kb),
	}
	return list
}

// fingerprintMatcher is the minimal KB surface fingerprint_match needs.
type fingerprintMatcher interface {
	FingerprintMatch(component, errorPattern, version string) map[string]any
}

// ---------------------------------------------------------------------------
// Tool 1: query_yarn
// ---------------------------------------------------------------------------

func queryYarnTool(c cluster.Cluster) tools.Def {
	return tools.Def{
		Name:        "query_yarn",
		Description: "Query YARN RM. Use 'list_apps' to see all apps (FAILED first). Use 'app_nodes' with app_id to find nodes.",
		InputSchema: tools.Object(map[string]any{
			"action":   tools.Prop("'list_apps' or 'app_nodes'"),
			"app_id":   tools.Prop("YARN app ID for 'app_nodes'"),
			"app_name": tools.Prop("Comma-separated states: RUNNING,FAILED,KILLED"),
		}, []string{"action"}),
		Handler: func(ctx context.Context, args map[string]any) (tools.ToolResult, error) {
			action := tools.ArgStr(args, "action", "list_apps")
			appName := tools.ArgStr(args, "app_name", "RUNNING,FAILED,KILLED,FINISHED")
			appID := tools.ArgStr(args, "app_id", "")
			return queryYarnHandler(c, action, appID, appName), nil
		},
	}
}

func queryYarnHandler(c cluster.Cluster, action, appID, appName string) tools.ToolResult {
	// The simulated cluster has no YARN REST API; query_yarn in demo mode reads
	// the scenario's yarn data. Real mode would hit rm_url. Keep a faithful
	// placeholder that returns node info from the cluster context.
	if action == "list_apps" {
		return tools.OK("0 YARN apps found (simulated).")
	}
	if action == "app_nodes" {
		return tools.OK(fmt.Sprintf("App %s node info unavailable (simulated).", appID))
	}
	return tools.OK(fmt.Sprintf("Unknown action: %s", action))
}

// ---------------------------------------------------------------------------
// Tool 2: ssh_exec (with read-only command safety)
// ---------------------------------------------------------------------------

func sshExecTool(c cluster.Cluster) tools.Def {
	return tools.Def{
		Name:        "ssh_exec",
		Description: "Execute read-only shell commands on cluster nodes via SSH. Find logs, grep errors, check processes.",
		InputSchema: tools.Object(map[string]any{
			"node_name": tools.Prop("Node: 'master1', 'core1', or full node_name"),
			"cmd":       tools.Prop("Shell command: find, grep, tail, ps, curl, df, free"),
			"timeout_s": tools.IntProp("SSH timeout in seconds"),
		}, []string{"node_name", "cmd"}),
		Handler: func(ctx context.Context, args map[string]any) (tools.ToolResult, error) {
			nodeName := tools.ArgStr(args, "node_name", "")
			cmd := tools.ArgStr(args, "cmd", "")
			return sshExecHandler(c, nodeName, cmd), nil
		},
	}
}

func sshExecHandler(c cluster.Cluster, nodeName, cmd string) tools.ToolResult {
	if allowed, reason := ValidateCommand(cmd); !allowed {
		return tools.OK(fmt.Sprintf("Command blocked by safety filter: %s", reason))
	}
	// Simulated cluster returns an empty result (no real SSH in demo). The real
	// adapter would exec over paramiko-equivalent SSH.
	_ = nodeName
	return tools.OK("(simulated) command validated but not executed")
}

// ValidateCommand checks that a shell command is read-only and safe.
// Fail-closed: every pipeline/compound segment must start with a command from
// the strict read-only allowlist, output redirection is rejected outright, and
// known write-capable argument forms (-delete, -exec, sed -i, ...) are blocked.
func ValidateCommand(cmd string) (bool, string) {
	clean := strings.TrimSpace(cmd)
	if clean == "" {
		return false, "empty command"
	}
	if strings.ContainsAny(clean, "\n\r") {
		return false, "multi-line commands not allowed"
	}
	lower := strings.ToLower(clean)

	// Redirection / substitution / expansion always rejected.
	for _, f := range []string{">", "<", "`", "$(", "${", "\\", "\x00"} {
		if strings.Contains(clean, f) {
			return false, fmt.Sprintf("Forbidden pattern '%s' in command", f)
		}
	}

	// Split on compound operators and require each segment to be allowlisted.
	segs := splitSegments(lower)
	for _, seg := range segs {
		fields := strings.Fields(seg)
		if len(fields) == 0 {
			continue
		}
		if !readOnlyAllow[fields[0]] {
			return false, fmt.Sprintf("Command '%s' not in read-only allowlist", fields[0])
		}
	}

	// Argument-level write primitives.
	for _, tok := range []string{"-delete", "-exec", "-execdir", "-ok", "-okdir", "chmod", "chown", "kill"} {
		if containsAnyToken(lower, tok) {
			return false, fmt.Sprintf("Forbidden argument '%s' in command", tok)
		}
	}
	return true, ""
}

// readOnlyAllow is the strict read-only command allowlist. Note the deliberate
// absence of sed/awk/echo/curl: sed -i and awk 'print > "f"' can write files,
// echo can emit arbitrary content, and curl can upload data out of the host.
var readOnlyAllow = map[string]bool{
	"cat": true, "grep": true, "egrep": true, "fgrep": true,
	"find": true, "tail": true, "head": true, "less": true, "more": true,
	"ls": true, "stat": true, "du": true, "df": true, "free": true,
	"ps": true, "top": true, "uptime": true, "uname": true, "hostname": true,
	"whoami": true, "id": true, "date": true, "pwd": true,
	"ss": true, "netstat": true, "ip": true,
	"sort": true, "uniq": true, "cut": true, "tr": true, "wc": true,
	"jq": true, "column": true,
	"yarn": true, "hdfs": true, "zkcli": true,
}

// splitSegments splits a command on ; && || and | (quote-naively — fail-closed:
// a quoted | yields an extra segment that must itself be allowlisted).
func splitSegments(cmd string) []string {
	s := strings.ReplaceAll(cmd, "&&", ";")
	s = strings.ReplaceAll(s, "||", ";")
	return strings.FieldsFunc(s, func(r rune) bool { return r == ';' || r == '|' })
}

func containsAnyToken(s, tok string) bool {
	for _, f := range strings.Fields(s) {
		if f == tok {
			return true
		}
	}
	return false
}

// ---------------------------------------------------------------------------
// Tool 3: deepwiki_query
// ---------------------------------------------------------------------------

func deepwikiTool() tools.Def {
	repoMap := map[string]string{
		"flink": "apache/flink", "hdfs": "apache/hadoop", "yarn": "apache/hadoop",
		"hadoop": "apache/hadoop", "kafka": "apache/kafka", "spark": "apache/spark",
		"hbase": "apache/hbase", "hive": "apache/hive",
	}
	return tools.Def{
		Name:        "deepwiki_query",
		Description: "Query known issues in open-source repos. Component→repo: flink→apache/flink, hdfs→apache/hadoop, kafka→apache/kafka.",
		InputSchema: tools.Object(map[string]any{
			"component": tools.Prop("Component: flink, hdfs, yarn, kafka, spark, hbase"),
			"question":  tools.Prop("Natural language question about known bugs"),
			"version":   tools.Prop("Component version"),
		}, []string{"component", "question"}),
		Handler: func(ctx context.Context, args map[string]any) (tools.ToolResult, error) {
			component := tools.ArgStr(args, "component", "")
			question := tools.ArgStr(args, "question", "")
			version := tools.ArgStr(args, "version", "")
			return deepwikiHandler(repoMap, component, question, version), nil
		},
	}
}

func deepwikiHandler(repoMap map[string]string, component, question, version string) tools.ToolResult {
	repo, ok := repoMap[strings.ToLower(component)]
	if !ok {
		return tools.OK(fmt.Sprintf("Unknown component '%s'.", component))
	}
	// MCP call to mcp.deepwiki.com would go here in production. In demo/offline
	// mode we return a placeholder carrying the structured query.
	q := question
	if version != "" {
		q = fmt.Sprintf("%s (version: %s)", question, version)
	}
	return tools.OK(fmt.Sprintf("[%s] deepwiki query: %s (not executed in demo)", repo, q))
}

// ---------------------------------------------------------------------------
// Tool 4: fingerprint_match
// ---------------------------------------------------------------------------

func fingerprintTool(kb fingerprintMatcher) tools.Def {
	return tools.Def{
		Name:        "fingerprint_match",
		Description: "Check if this error matches a known case. Fast-path before full diagnosis.",
		InputSchema: tools.Object(map[string]any{
			"component":     tools.Prop("Component: flink, hdfs, yarn"),
			"error_pattern": tools.Prop("Key error: OutOfMemoryError, CheckpointExpired..."),
			"version":       tools.Prop("Component version"),
		}, []string{"component", "error_pattern"}),
		Handler: func(ctx context.Context, args map[string]any) (tools.ToolResult, error) {
			component := tools.ArgStr(args, "component", "")
			errorPattern := tools.ArgStr(args, "error_pattern", "")
			version := tools.ArgStr(args, "version", "")
			if kb != nil {
				if hit := kb.FingerprintMatch(component, errorPattern, version); hit != nil {
					return tools.OK(fmt.Sprintf("Known issue (confidence: high):\n  Root cause: %v\n  Suggestions: %v", hit["root_cause"], hit["suggestions"])), nil
				}
			}
			return tools.OK(fmt.Sprintf("No fingerprint match for %s/%s. Proceed with diagnosis.", component, errorPattern)), nil
		},
	}
}

// ---------------------------------------------------------------------------
// Tool 5: call_umr_agent (HMAC-SHA1 signed)
// ---------------------------------------------------------------------------

func umrAgentTool(c cluster.Cluster) tools.Def {
	return tools.Def{
		Name:        "call_umr_agent",
		Description: "Call umrAgent on cluster nodes. Actions: GetLogs(Path,Keywords,MaxLines,Since), CheckProcess(ProcessName), GetBaseInfo, GetAppList.",
		InputSchema: tools.Object(map[string]any{
			"node_name": tools.Prop("Node: 'master1', 'core1', or full node_name"),
			"action":    tools.Prop("umrAgent Action: GetLogs, CheckProcess, GetBaseInfo, GetAppList"),
			"params":    tools.PropObject("Action-specific params"),
		}, []string{"node_name", "action"}),
		Handler: func(ctx context.Context, args map[string]any) (tools.ToolResult, error) {
			nodeName := tools.ArgStr(args, "node_name", "")
			action := tools.ArgStr(args, "action", "")
			var params map[string]any
			if p, ok := args["params"].(map[string]any); ok {
				params = p
			}
			return umrAgentHandler(c, nodeName, action, params), nil
		},
	}
}

func umrAgentHandler(c cluster.Cluster, nodeName, action string, params map[string]any) tools.ToolResult {
	node := c.FindNode(nodeName)
	if node == nil {
		return tools.OK(fmt.Sprintf("Error: Node '%s' not found", nodeName))
	}
	// Simulated cluster carries log content; action GetLogs/GetBaseInfo map to
	// reading scenario logs. For demo, return a placeholder.
	_ = node
	return tools.OK(fmt.Sprintf("umrAgent %s on %s (simulated)", action, nodeName))
}

// ---------------------------------------------------------------------------
// HMAC-SHA1 signing (byte-identical to util/agent.js + pkg/client/uagent)
// ---------------------------------------------------------------------------

// SignParams computes HMAC-SHA1 over sorted-key values, identical to the
// Node.js util/agent.js auth() and Go pkg/client/uagent setGetSignature().
func SignParams(params map[string]string, key string) string {
	keys := make([]string, 0, len(params))
	for k := range params {
		keys = append(keys, k)
	}
	sort.Strings(keys)

	var b strings.Builder
	for _, k := range keys {
		b.WriteString(params[k])
	}
	mac := hmac.New(sha1.New, []byte(key))
	mac.Write([]byte(b.String()))
	return hex.EncodeToString(mac.Sum(nil))
}

// DateMS returns the current UnixMilli timestamp as a string (for signing).
func DateMS() string {
	return fmt.Sprintf("%d", time.Now().UnixMilli())
}

// MD5Fingerprint computes md5(component:error:version)[:16].
func MD5Fingerprint(component, errorPattern, version string) string {
	sum := md5.Sum([]byte(fmt.Sprintf("%s:%s:%s", component, errorPattern, version)))
	return hex.EncodeToString(sum[:])[:16]
}

func contains(list []string, s string) bool {
	for _, x := range list {
		if x == s {
			return true
		}
	}
	return false
}
